internal/tasks/device_render.go

```diff
@@ -0,0 +1,173 @@
+package tasks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	config_latest "github.com/coreos/ignition/v2/config/v3_4"
+	config_latest_types "github.com/coreos/ignition/v2/config/v3_4/types"
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/store"
+	"github.com/flightctl/flightctl/internal/util"
+	"github.com/flightctl/flightctl/pkg/log"
+	"github.com/flightctl/flightctl/pkg/reqid"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+func DeviceRender(taskManager TaskManager) {
+	for {
+		select {
+		case <-taskManager.ctx.Done():
+			taskManager.log.Info("Received ctx.Done(), stopping")
+			return
+		case resourceRef := <-taskManager.channels[ChannelDeviceRender]:
+			requestID := reqid.NextRequestID()
+			ctx := context.WithValue(context.Background(), middleware.RequestIDKey, requestID)
+			log := log.WithReqIDFromCtx(ctx, taskManager.log)
+			logic := NewDeviceRenderLogic(taskManager, log, taskManager.store, resourceRef)
+			if resourceRef.Op == DeviceRenderOpUpdate {
+				err := logic.RenderDevice(ctx)
+				if err != nil {
+					log.Errorf("failed rendering device %s/%s: %v", resourceRef.OrgID, resourceRef.Name, err)
+				}
+			} else {
+				log.Errorf("DeviceRender called with unexpected kind %s and op %s", resourceRef.Kind, resourceRef.Op)
+			}
+		}
+	}
+}
+
+type DeviceRenderLogic struct {
+	taskManager    TaskManager
+	log            logrus.FieldLogger
+	store          store.Store
+	resourceRef    ResourceReference
+	device         *api.Device
+	renderedConfig *config_latest_types.Config
+}
+
+func NewDeviceRenderLogic(taskManager TaskManager, log logrus.FieldLogger, store store.Store, resourceRef ResourceReference) DeviceRenderLogic {
+	return DeviceRenderLogic{taskManager: taskManager, log: log, store: store, resourceRef: resourceRef}
+}
+
+func (t *DeviceRenderLogic) RenderDevice(ctx context.Context) error {
+	t.log.Infof("Rendering device %s/%s", t.resourceRef.OrgID, t.resourceRef.Name)
+
+	device, err := t.store.Device().Get(ctx, t.resourceRef.OrgID, t.resourceRef.Name)
+	if err != nil {
+		return err
+	}
+	t.device = device
+	emptyIgnitionConfig, _, _ := config_latest.ParseCompatibleVersion([]byte("{\"ignition\": {\"version\": \"3.4.0\"}"))
```

Comment:
question: should the supported IGN version be a const somewhere?

Reply:
Do you mean what version the agent supports?  Good question.  My hunch is:
1. We need version negotiation for the entire agent-service API or at least some mechanism for the agent to upgrade itself to the latest version.
2. The ignition version we use would be tied to the overall version of the API.

WDYT?

Reply:
Yes, a protocol version makes sense but for now, we should put the ign version `3.4` as a public const vs copying it around.